### ISSUE TITLE
makecmds in non-linux systems

### ISSUE THREAD

**Issue by cenk1cenk2** [loichyan/tmux-toggle-popup#12]
Sorry to disturb you again @loichyan,

Everything works on the linux side of things with me, but on the retarded macos side which I am forced to use at work, `xargs` causes me to interpret the `printf` command with the system one which does not have the `%q` interpolation, and breaks everything.

I have given this a try myself with no avail. The idea was to interpret the `printf` with the `builtin` `bash` command so that it does not use the system one.

https://github.com/loichyan/tmux-toggle-popup/blob/main/scripts/helpers.sh#L41

This is close but still not there since it really not escapes nor performs properly with having extra quotes.

```bash
# Parses the tmux script into sequences and escapes each one, ensuring they can
# be safely interpreted by Bash.
makecmds() {
	xargs -0 -I {} bash -c 'printf "%q " $*' _ {}
}
```

Any advise would be great if you have any expertise on this one.

**Comment by loichyan** [comment]
> I have given this a try myself with no avail. The idea was to interpret the `printf` with the `builtin` `bash` command so that it does not use the system one.

I initially thought the bash built-in `printf` would always be selected due to the presence of `#!/usr/bin/env bash` shebang in each file, but that doesn't seem to be the case.

> This is close but still not there since it really not escapes nor performs properly with having extra quotes.

I don't have a macOS machine, but fortunately, we can use the macOS image from GitHub Actions. I'm currently working on setting up CI/CD tests and will try to fix this issue afterward :)

**Comment by cenk1cenk2** [comment]
I can give it a try if you give me some directions on how to perform this either with xargs or without it completely. Since it seems I am also not prepared in this weird cases a mac has.

**Comment by loichyan** [comment]
> I can give it a try if you give me some directions on how to perform this either with xargs or without it completely. Since it seems I am also not prepared in this weird cases a mac has.

The parsing process is generally as follows:

1. First, we use `xargs` to parse `display "yyy ppp"` into two arguments: `display` and `yyy ppp`. 
2. Then, `printf "%q "` will add some escaping to prevent each argument from being split by Bash. For instance, it may escape them as `display` and `yyy\ ppp`. 
3. Finally, the result is executed directly with `tmux` using `eval`, which, in this case, would be `eval tmux display yyy\ ppp`. 

You can find some parsing tests in [tests/commands-parser.sh](https://github.com/loichyan/tmux-toggle-popup/blob/78cd5531614fc1134186012381856c2bb4126eeb/tests/commands-parser.sh). If it passes on macOS, the issue should be resolved.

**Comment by loichyan** [comment]
It should be fixed in #13, as all tests pass on macOS.

Additionally, I would like to highlight a few mistakes in your fixes that you might find interesting :)

```bash
makecmds() {
    xargs -0 -I {} bash -c 'printf "%q " $*' _ {}
    #      ^  ^^^^                       ^^
    #      1)   2)                       3)
}
```

1. According to the `xargs` man page, the `-0` option causes it to treat each character literally. As a result, it does not parse escapes, quotes, etc.
2. The `-I {}` option tells `xargs` to concatenate all arguments into a single string and replace each occurrence of `{}` with that string.
3. We should use `"$@"` to prevent Bash from splitting the parsed arguments again.

**Comment by cenk1cenk2** [comment]
Makes perfect sense why it was failing in my case. Thank you for explanation.
